// sourcepolicyplugin/index.js
module.exports = function (app) {
  const received = []
  let unsubscribes = []
  return {
    id: 'sourcepolicyplugin',
    name: 'Source Policy Test Plugin',
    schema: function () {
      return {
        type: 'object',
        properties: {
          sourcePolicy: { type: 'string' }
        }
      }
    },
    start: function (options) {
      const command = {
        context: 'vessels.self',
        subscribe: [{ path: 'environment.inside.test.pressure' }]
      }
      if (options && options.sourcePolicy) {
        command.sourcePolicy = options.sourcePolicy
      }
      app.subscriptionmanager.subscribe(
        command,
        unsubscribes,
        (err) => console.error(err),
        (delta) => {
          delta.updates.forEach((u) => {
            if (!u.values) return
            u.values.forEach((v) => {
              if (v.path === 'environment.inside.test.pressure') {
                received.push({ $source: u.$source, value: v.value })
              }
            })
          })
        }
      )
    },
    stop: function () {
      unsubscribes.forEach((f) => f())
      unsubscribes = []
    },
    _received: received
  }
}
