// upgradetestplugin/index.js
const { WebSocketServer } = require('ws')

module.exports = function (app) {
  // Single WebSocketServer reused across plugin start/stop cycles. The
  // server-api contract says registrations made via registerWithUpgrade
  // persist for the lifetime of the SignalK process, so the underlying
  // WebSocket machinery should also be persistent. The `started` flag
  // is checked inside the handler to reject upgrades while stopped.
  const wss = new WebSocketServer({ noServer: true })
  let started = false

  return {
    id: 'upgradetestplugin',
    name: 'Upgrade Test Plugin',
    schema: { type: 'object', properties: {} },
    start: function () {
      started = true
    },
    stop: function () {
      started = false
    },
    registerWithUpgrade: function (upgrader) {
      upgrader.upgrade('/echo', (request, socket, head) => {
        if (!started) {
          socket.destroy()
          return
        }
        wss.handleUpgrade(request, socket, head, (ws) => {
          ws.on('message', (msg) => {
            ws.send('echo:' + msg.toString())
          })
        })
      })
    }
  }
}
