// testplugin/index.js
module.exports = function(app) {
  return {
    id: 'testplugin',
    name: 'Test Plugin Name',
    schema: function() {
      return {
        type: 'object',
        properties: {}
      }
    },
    start: function() {
      app.debug('Demoing debug logging')
      app.error('Demoing error diagnostics')
      this.started = true

      app.handleMessage(this.id, {
        updates: [
          {
            values: [
              {
                path: 'some.path',
                value: 'someValue'
              }
            ]
          }
        ]
      })

      app.registerDeltaInputHandler((delta, next) => {
        if (
          delta.updates[0].values[0].path === 'navigation.courseOverGroundTrue'
        ) {
          delta.updates[0].values[0].value = -1
        }
        next(delta)
      })

      app.registerResourceProvider({
        type: 'test-custom-resourcetype',
        methods: {
          listResources: (type, query) => {
            console.log(`test plugin listResources(${type}, ${query})`)
            return Promise.resolve({
              dummyResourceId: {
                value: `test plugin listResources(${type}, ${query})`
              }
            })
          },
          getResource: (type, id) => {
            console.log(`test plugin getResource(${type}, ${id})`)
            return Promise.resolve({
              value: `test plugin listResources(${type}, ${id})`
            })
          },
          setResource: (type, id, value) => {
            console.log(`test plugin setResource(${type}, ${id}, ${value})`)
            return Promise.resolve({
              dummyResourceId: {
                value: `test plugin listResources(${type}, ${id}, ${value})`
              }
            })
          },
          deleteResource: (type, id) => {
            console.log(`test plugin deleteResource(${type}, ${id})`)
            return Promise.resolve()
          }
        }
      })
    },
    stop: function() {
      this.started = false
    },
    registerWithRouter: (router) => {
      router.get('/demopluginGet', (req, res) => res.json({message: 'test ok'}))
      router.get('/readonlyData', (req, res) => res.json({data: 'readonly ok'}))
      router.post('/writeData', (req, res) => res.json({data: 'write ok'}))
      router.get('/sensorData/:id', (req, res) => res.json({sensor: req.params.id}))
    },
    getRoutePermissions: () => [
      { method: 'GET', path: '/readonlyData', permission: 'readonly' },
      { method: 'POST', path: '/writeData', permission: 'readwrite' },
      { method: 'GET', path: '/sensorData/:id', permission: 'readonly' }
    ],
    getOpenApi: () => require('./openApi'),
    app: app
  }
}
